### ISSUE TITLE
SIGKILL caused by Realm file locks in app vs NSE

### ISSUE THREAD

**Issue by IAmMichellis** [matrix-org/matrix-ios-sdk#1452]
**Describe the bug**

TL;DR: we believe that MatrixSDK is SIGKILL crashing when requesting Realm. We think it's happening because of aggressive system killing on holding file locks (on Realm) in a shared app container when the app goes to background and the Notification Service Extension (NSE) attempts to access the shared Realm database.

Longer version:
On Beeper (which forks this sdk) we're seeing TestFlight capturing SIGKILLs. There's no obvious reason for the crash, and the main thread is typically idle, but somewhere in the crash log we see something like this:

```

Exception Type:  EXC_CRASH (SIGKILL)
Exception Codes: 0x0000000000000000, 0x0000000000000000
Exception Note:  EXC_CORPSE_NOTIFY
Termination Reason: RUNNINGBOARD 3735883980 

Triggered by Thread:  0


Thread 0 name:
Thread 0 Crashed:
0   libsystem_kernel.dylib        	0x00000001fa03c4e0 mach_msg_trap + 8
1   libsystem_kernel.dylib        	0x00000001fa03cb24 mach_msg + 76 (mach_msg.c:119)
2   CoreFoundation                	0x00000001c24e7820 __CFRunLoopServiceMachPort + 372 (CFRunLoop.c:2646)
3   CoreFoundation                	0x00000001c24ebcac __CFRunLoopRun + 1180 (CFRunLoop.c:3000)
4   CoreFoundation                	0x00000001c24ff6b8 CFRunLoopRunSpecific + 600 (CFRunLoop.c:3268)
5   GraphicsServices              	0x00000001de599374 GSEventRunModal + 164 (GSEvent.c:2200)
6   UIKitCore                     	0x00000001c4e64e88 -[UIApplication _run] + 1100 (UIApplication.m:3511)
7   UIKitCore                     	0x00000001c4be65ec UIApplicationMain + 364 (UIApplication.m:5064)
8   Riot                          	0x0000000102dd54b0 main + 68 (AppDelegate.swift:21)
9   dyld                          	0x0000000104a39ce4 start + 520 (dyldMain.cpp:879)

.
.
.

Thread 16 name:
Thread 16:
0   libsystem_kernel.dylib        	0x00000001fa03dd6c __open + 8
1   libsystem_kernel.dylib        	0x00000001fa03e150 open + 40 (open-base.c:101)
2   Realm                         	0x0000000106361064 0x106084000 + 3002468
3   Realm                         	0x00000001061dc7ec 0x106084000 + 1411052
4   Realm                         	0x0000000106213a68 0x106084000 + 1636968
5   Realm                         	0x0000000106216d24 0x106084000 + 1649956
6   Realm                         	0x000000010621c368 0x106084000 + 1672040
7   Realm                         	0x0000000106434d24 0x106084000 + 3869988
8   Realm                         	0x0000000106435ca0 0x106084000 + 3873952
9   Realm                         	0x0000000106435ac4 0x106084000 + 3873476
10  Realm                         	0x00000001064c8ad4 0x106084000 + 4475604
11  Realm                         	0x00000001061b2d3c +[RLMRealm realmWithConfiguration:queue:error:] + 956 (RLMRealm.mm:476)
12  Realm                         	0x00000001061b2970 +[RLMRealm realmWithConfiguration:error:] + 28 (RLMRealm.mm:398)
13  MatrixSDK                     	0x0000000105907ecc +[MXRealmCryptoStore realmForUser:andDevice:readOnlySnapshot:] + 2724 (MXRealmCryptoStore.m:1819)
14  MatrixSDK                     	0x0000000105902308 0x1057c4000 + 1303304
15  MatrixSDK                     	0x00000001058e8a0c 0x1057c4000 + 1198604
16  MatrixSDK                     	0x000000010581d168 0x1057c4000 + 364904
17  libdispatch.dylib             	0x00000001c21dee68 _dispatch_call_block_and_release + 32 (init.c:1517)
18  libdispatch.dylib             	0x00000001c21e0a2c _dispatch_client_callout + 20 (object.m:560)
19  libdispatch.dylib             	0x00000001c21e8124 _dispatch_lane_serial_drain + 668 (inline_internal.h:2622)
20  libdispatch.dylib             	0x00000001c21e8c80 _dispatch_lane_invoke + 392 (queue.c:3944)
21  libdispatch.dylib             	0x00000001c21f3500 _dispatch_workloop_worker_thread + 648 (queue.c:6732)
22  libsystem_pthread.dylib       	0x0000000233cae0bc _pthread_wqthread + 288 (pthread.c:2599)
23  libsystem_pthread.dylib       	0x0000000233cade5c start_wqthread + 8
```

This is an interesting stack trace because it shows that MatrixSDK was asking Realm for a Realm instance. That seems like a benign request. Except that we see this in crashes often, so it smells.

We suspect the issue is related to comments in this thread: https://githubhot.com/repo/realm/realm-swift/issues/7649

Specifically: we have an app and an NSE accessing the same Realm database (via MatrixSDK's MXRealmCryptoStore), and we believe that's causing the OS to kill us due to a file lock.

**To Reproduce**

Unknown: the app runs in background and the TestFlight "the app crashed" dialog appears. The logs look as described above.

**Expected behavior**
Fewer / no SIGKILL crashes with that Realm trace.

**Additional Context**
My colleague started a thread about this with Realm developers many months ago, but didn't get much traction: https://github.com/realm/realm-swift/issues/7466

I believe the thread mentioned in the main description has the key information, about shared app containers and filelocks: https://githubhot.com/repo/realm/realm-swift/issues/7649

**Comment by IAmMichellis** [comment]
Solved :)  Info / solution here https://github.com/realm/realm-swift/issues/7466#issuecomment-1149948784

Probably I should upstream changes.... I'll get on that.
